File: Utilities/Matlab/scripts/mass_balance_gas_volume.m
% McDermott
% 2 Dec 2021
% mass_balance_gas_volume.m

close all
clear all

plot_mass_balance('mass_balance_gas_volume','');

function [] = plot_mass_balance(chid,title_text)

plot_style
figure
set(gca,'Units',Plot_Units)
set(gca,'Position',[Plot_X Plot_Y Plot_Width Plot_Height])

ddir = '../../Verification/Species/';
M = importdata([ddir,chid,'_devc.csv'],',',2);

t = M.data(:,1);
m = M.data(:,find(strcmp(M.colheaders,'"M"')));
dmdt = zeros(length(t),1);
for i=2:length(t)
    dmdt(i) = (m(i)-m(i-1))/(t(i)-t(i-1));
end

mf_x1  = M.data(:,find(strcmp(M.colheaders,'"TMF_X1"')));
mf_x2  = M.data(:,find(strcmp(M.colheaders,'"TMF_X2"')));
mf_y1  = M.data(:,find(strcmp(M.colheaders,'"TMF_Y1"')));
mf_y2  = M.data(:,find(strcmp(M.colheaders,'"TMF_Y2"')));
mf_z1  = M.data(:,find(strcmp(M.colheaders,'"TMF_Z1"')));
mf_z2  = M.data(:,find(strcmp(M.colheaders,'"TMF_Z2"')));

bal = dmdt + (mf_x2-mf_x1 + mf_y2-mf_y1 + mf_z2-mf_z1);

plot(t,zeros(1,length(t)),'k-'); hold on
H(1)=plot(t,dmdt);
H(2)=plot(t,(mf_x2-mf_x1 + mf_y2-mf_y1 + mf_z2-mf_z1));
H(3)=plot(t,bal);

ylabel('mass flow (kg/s)', 'FontSize',Label_Font_Size)
xlabel('time (s)', 'FontSize',Label_Font_Size)

set(gca,'FontName',Font_Name)
set(gca,'FontSize',Label_Font_Size)

lh=legend(H,'dm/dt','out-in','dm/dt+out-in');
set(lh,'FontSize',Key_Font_Size)

text(100,18e-3,title_text,'FontSize',Title_Font_Size,'FontName',Font_Name,'Interpreter',Font_Interpreter)

% check balance and report error

mass_error = max(abs(bal));
if mass_error > 1e-6
    disp(['Matlab Warning: mass error = ',num2str(mass_error),' in ',chid])
end

% add version string if file is available

Git_Filename = [ddir,chid,'_git.txt'];
addverstr(gca,Git_Filename,'linear')

set(gcf,'Visible',Figure_Visibility);
set(gcf,'Units',Paper_Units);
set(gcf,'PaperUnits',Paper_Units);
set(gcf,'PaperSize',[Paper_Width Paper_Height]);
set(gcf,'Position',[0 0 Paper_Width Paper_Height]);
print(gcf,'-dpdf',['../../Manuals/FDS_Verification_Guide/SCRIPT_FIGURES/',chid])

end

File: Utilities/Python/scripts/mass_balance_gas_volume.py
"""
mass_balance_gas_volume.py
Validates mass conservation in a gas volume comparing time derivative of mass with net mass flux through boundaries.
Original MATLAB script by McDermott (2 Dec 2021)
"""
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import fdsplotlib
import os

def plot_mass_balance(chid, title_text):
    
    plot_style = fdsplotlib.get_plot_style('fds')
    fds_dir    = os.path.normpath(os.path.join(os.path.dirname(__file__),'..','..','..'))
    ddir       = os.path.join(fds_dir, 'Verification','Species','')
    pltdir     = os.path.join(fds_dir, 'Manuals','FDS_Verification_Guide','SCRIPT_FIGURES','')
    devc_file  = os.path.join(ddir, f'{chid}_devc.csv')

    try:
        M = pd.read_csv(devc_file, skiprows=1)  # Skip units row, use column names as header
    except FileNotFoundError:
        print(f'Error: File {devc_file} does not exist. Skipping case.')
        return 0

    t = M.iloc[:, 0].values
    m = M['M'].values
    dmdt = np.zeros(len(t))
    for i in range(1, len(t)):
        dmdt[i] = (m[i] - m[i-1]) / (t[i] - t[i-1])
    
    mf_x1 = M['TMF_X1'].values
    mf_x2 = M['TMF_X2'].values
    mf_y1 = M['TMF_Y1'].values
    mf_y2 = M['TMF_Y2'].values
    mf_z1 = M['TMF_Z1'].values
    mf_z2 = M['TMF_Z2'].values    
    # Net flux: out - in
    net_flux = mf_x2 - mf_x1 + mf_y2 - mf_y1 + mf_z2 - mf_z1
    # Residual: dm/dt + (out - in) (should be ~0)
    bal = dmdt + net_flux
        
    # Get version string if git file exists
    git_file = os.path.join(ddir, f'{chid}_git.txt')
    version_string = fdsplotlib.get_version_string(git_file) if os.path.exists(git_file) else ''
    
    # Plot zero reference line first (black solid line, no label)
    fig = fdsplotlib.plot_to_fig(x_data=t, y_data=np.zeros_like(t), marker_style='k-',
                                 revision_label=version_string, legend_location='upper right',x_label='time (s)', y_label='mass flow (kg/s)',
                                 x_min=0, x_max=2000, y_min=-0.004, y_max=0.004,xnumticks=5, ynumticks=5)
    fdsplotlib.plot_to_fig(x_data=t, y_data=dmdt, figure_handle=fig, marker_style='r-', data_label='dm/dt')
    fdsplotlib.plot_to_fig(x_data=t, y_data=net_flux, figure_handle=fig, marker_style='m-', data_label='out-in')
    fdsplotlib.plot_to_fig(x_data=t, y_data=bal, figure_handle=fig, marker_style='b-', data_label='dm/dt+out-in')
    
    ax = plt.gca()
    lines = ax.get_lines()
    lines[2].set_color('orange')  # Orange for net flux (second data line)
    legend = ax.get_legend()
    if legend:
        legend.legend_handles[1].set_color('orange')  # Second legend entry (out-in)
    fdsplotlib.apply_global_exponent(ax, axis='y')
    if title_text:
        ax.text(100, 18e-3, title_text, fontsize=plot_style['Title_Font_Size'], 
                fontname=plot_style['Font_Name'])
    
    # Check balance and report error
    mass_error = np.max(np.abs(bal))
    if mass_error > 1e-6:
        print(f'Python Warning: mass error = {mass_error:.6e} in {chid}')
    
    # Save figure
    output_file = os.path.join(pltdir, f'{chid}.pdf')
    plt.savefig(output_file, format='pdf')
    plt.close(fig)
        
    return mass_error

if __name__ == '__main__':
    """Main execution block"""
    print("Running mass balance gas volume verification plot...")
    error = plot_mass_balance('mass_balance_gas_volume', '')
    print("Mass balance gas volume verification completed successfully!")
